text/0517-io-os-reform.md

```diff
@@ -556,6 +556,15 @@ buffers as input. This has multiple benefits:
   results are wanted, one can use `read_to_end` and convert to a
   `String` only at the end.
 
+`read_to_end` and `read_to_string` return a `Result` with either the
+number of bytes read or an error.  If a short read occurs, that call
```

Comment:
> In this context, what would you define as a "short read"? If you're asking for the entire
> contents of the stream then a "short read" implies you already know how many bytes are in
> the stream in the first place.

We talked about this above, I conceded the point.  But to reiterate, the intent was to allow the implementation not to implement a busy loop if given a non-blocking source.

> Additionally, below you mention a "non-transient" error, but how do we classify errors as
> transient or not transient?

Not sure it's worth discussing, given the above, but: the intent was to give liberty to an implementation to just return whatever it got from the syscall.

-----

Anyway, we've established that this solution sucks, but I still believe there should be *something* that prevents these from getting called with non-blocking sources, because it can't have any reasonable semantics for that case.